### ISSUE TITLE
latest etcd client tries to load /v2/members rather than /v2/machines

### ISSUE THREAD

**Issue by lordelph** [kelseyhightower/confd#195]
Running a new build of confd against an existing etcd 0.4.6 it fails to identify the cluster even though it is reachable with curl. confd returns the error

    ERROR cannot connect to etcd cluster: 172.17.8.102:4001

But `curl http://172.17.8.102:4001/v2/machines` succeeds.

I pulled down the confd code and traced through it to discover what was happening. It looks like the etcd client now looks for /v2/members as of 3 days ago - [see this commit](https://github.com/coreos/go-etcd/commit/5348c80bde6223fa53c48ef546b4802db69b1d70#diff-33c1223aac4ef741731cc8c73b4d7777L295)

Of course, when run against an older etcd, this fails and produces a 404, and confd fails.

I think this is a robustness problem for etcd rather than confd, but I thought it worth logging here in case others start experiencing it.

**Comment by bacongobbler** [comment]
Since they're going to create a fork of the etcd client which supports 0.4.6 (according to their comments), we should probably support both `etcd` and `etcd2` drivers, with the `etcd` driver using the fork of the go-etcd client, and implementing `etcd2

**Comment by hkjn** [comment]
I was also hit by this, and also thought the issue was with etcd or go-etcd: https://github.com/coreos/etcd/issues/2273

Support for both major versions of etcd within confd would be nice, since currently the tutorial at https://www.digitalocean.com/community/tutorials/how-to-use-confd-and-etcd-to-dynamically-reconfigure-services-in-coreos won't work on CoreOS, which has etcd v0.4.6.

In the interrim, what's a clean solution to producing a `confd` binary that uses the 0.4.6-tagged branch of go-etcd? Needing to `git clone [go-etcd repo] && git checkout v0.4.6` in a Dockerfile (in addition to mucking around with godeps) to build `confd`, seems a bit messy..

**Comment by bacongobbler** [comment]
The [vendored code](https://github.com/kelseyhightower/confd/blob/master/Godeps/Godeps.json#L14-L16) via Godep will have to be updated to use the 0.4.6 branch. That should suffice for the time being.

**Comment by lordelph** [comment]
@hkjn That tutorial should still be valid, as it advises using a binary release. I'm using the latest binary v.0.7.1 successfully against CoreOS stable 557.2.0. Typical Dockerfile entry to install it:

    RUN apt-get update && \
        DEBIAN_FRONTEND=noninteractive apt-get -y install curl && \
        curl -o /usr/bin/confd -L https://github.com/kelseyhightower/confd/releases/download/v0.7.1/confd-0.7.1-linux-amd64 && \
        chmod 755 /usr/bin/confd

What's the best way to plan for when CoreOS updates to etcd 2.0 though? Suddenly, at reboot, confd 0.7.1 will stop working. Assuming there won't be a one-size-fits-all confd binary, I'm guessing I'll need a container which has two confd binaries and inspects etcd to decide which one to run?

**Comment by bacongobbler** [comment]
> What's the best way to plan for when CoreOS updates to etcd 2.0 though? Suddenly, at reboot, confd 0.7.1 will stop working.

Unfortunately I don't see a way around that. The go-etcd library locked for 0.7.1 is strictly compatible with 0.4.6 and earlier etcd clusters. etcd2 is backwards incompatible, so there's not much we can do there. :neutral_face: 

> Assuming there won't be a one-size-fits-all confd binary, I'm guessing I'll need a container which has two confd binaries and inspects etcd to decide which one to run?

We could shoehorn support for older and newer etcd clusters in confd. A solution could be to implement an `etcd2` backend, which will be more or less the exact same as the current etcd backend. The only change required will be that it'll be using the updated go-etcd client libs. That way you can use `confd --backend etcd2` for etcd2 clusters. For your use case, you'd still need to inspect etcd to determine whether you'd need to use either the etcd or the etcd2 backend, though.

Of course, the other solution is to drop etcd 0.4.6 support in confd 0.8, but that decision's up to @kelseyhightower. :)

**Comment by hkjn** [comment]
@lordelph That's what's odd here; I was indeed following the tutorial and downloading the same v.0.7.1 release binary, but hit this issue in CoreOS 522.6.0.

My Dockerfile has:
RUN apt-get update && apt-get -y install nginx curl
RUN curl -L https://github.com/kelseyhightower/confd/releases/download/v0.7.1/confd-0.7.1-linux-amd64 -o /usr/local/bin/confd
RUN chmod +x /usr/local/bin/confd

Any suggestions for what differs here would be welcome..

**Comment by hkjn** [comment]
Okay, I've figured out what was causing confd to behave against my expectations in my case:
1. When confd is started I pass it the docker0 interface's IP, to allow the container to reach etcd on the CoreOS host, thusly: `confd -node http://172.17.42.1:4001 ...`.
2. [backends/etcd/client.go's `NewEtcdClient()`](https://github.com/kelseyhightower/confd/blob/master/backends/etcd/client.go#L31) calls `c.SetCluster()` in go-etcd, which (in the correct v0.4.6 tag of the package) [ends up](https://github.com/coreos/go-etcd/blob/v0.4.6/etcd/client.go#L301) doing `GET http://172.17.42.1:4001/v2/machines` to determine the members in the etcd cluster.
3. That etcd endpoint returns the loopback IP, at least on my single-member CoreOS cluster:
```
# curl 172.17.42.1:4001/v2/machines; echo
http://127.0.0.1:4001
```
4. Thus, the go-etcd `Client` used by confd ends up with a set of machines that cannot be reached by it from inside the container. 

So as far as I can follow, maybe there's nothing confd could/should be doing differently, but the outcome sure was confusing to me, at least. The one suggestions I would make would be to see if you'd consider including the peers in the logged error message (available via go-etcd's `Client.GetCluster()`), since that likely would have led me down the right path faster.

Now to go find out why my etcd thinks loopback is a reasonable IP to have (on a GCE test cluster, the same `curl` command shows the public IPs of the VMs, which seems saner)..

**Comment by bacongobbler** [comment]
@hkjn if you nail it down, PRs/issues are always welcome! If you are still having trouble let's have this discussion on another bug since it seems like a separate issue than what @lordelph originally described in the OP.

**Comment by hkjn** [comment]
You're right @bacongobbler, I dropped and pruned the comments that were irrelevant to the issue. PRs may come for the logging changes I suggested, sorry for the noise!

**Comment by kelseyhightower** [comment]
This issue seems to be affecting many people right now.

**Comment by kelseyhightower** [comment]
Fixed on master and etcd 0.8.0 has been released: https://github.com/kelseyhightower/confd/releases/tag/v0.8.0
